lock-api/src/main/java/com/palantir/lock/watch/LockWatchEventCache.java

```diff
@@ -0,0 +1,59 @@
+/*
+ * (c) Copyright 2020 Palantir Technologies Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.lock.watch;
+
+import java.util.Collection;
+import java.util.Optional;
+import java.util.Set;
+
+import com.palantir.lock.v2.LockToken;
+
+public interface LockWatchEventCache {
+    /**
+     * Returns the last known lock watch version for the cache.
+     */
+    Optional<IdentifiedVersion> lastKnownVersion();
+
+    /**
+     * Updates the cache with the update, and identifies the given timestamps with that lock watch state.
+     */
+    void processTransactionUpdate(Collection<Long> startTimestamps, LockWatchStateUpdate update);
+
+    /**
+     * Updates the cache with the update, and calculates the {@link CommitUpdate} taking into account all changes to
+     * lock watch state since the start of the transaction, excluding the transaction's own commit locks.
+     *
+     * @param startTs          start timestamp of the transaction
+     * @param commitTs         commit timestamp of the transaction
+     * @param commitLocksToken lock token for the transactions's commit locks
+     * @return the commit update for this transaction's precommit condition
+     */
+    CommitUpdate getCommitUpdate(long startTs, long commitTs, LockToken commitLocksToken);
```

Comment:
Maybe I'm missing something here, but I don't see how it is possible to have sufficient information to exclude a set of locks based solely on the `LockToken`. At the time that this is taken out in the commit flow, we do have access to `Set<LockDescriptor>` which we register; I wonder if it is worth changing this interface to use that instead - it would certainly be easier and potentially just the right thing to do, but open to suggestions.

Reply:
Should be possible: each LockEvent/UnlockEvent has the LockToken; just need to figure out what the right id for filtering is?

Reply:
aaah
`LockWatchEvent` obscures things by a layer, I missed that. Good catch!

Reply:
Unlock events don't have a lock token, but then we don't strictly care about unlock events in this update at all; we only want the set of lock descriptors (excluding commit locks) that have been taken out at some point in this transaction; if lock X was locked then unlocked, we clearly still want to return X (i.e. the unlock events are irrelevant for the commit update / conflict checking of cached reads).